core/src/main/java/org/commonjava/indy/core/change/StoreContentListener.java

```diff
@@ -236,89 +251,130 @@ private void removeAllSupercededMemberContent( final ArtifactStore store,
+                            clearPathProcessor( p, memberKey, groups ) ) );
 
-                    clearPaths( paths, memberKey, groups, false );
                 } );
 
+                drainAndCount( clearService, "store: " + store.getKey() );
             }
         }
     }
 
-    private void clearPaths( Set<String> paths, StoreKey originKey, Set<Group> affectedGroups, boolean deleteOriginPath )
+    private Callable<Integer> clearPathProcessor( final String path, final StoreKey key, final Set<Group> groups )
+    {
+        try
+        {
+            ArtifactStore store = storeDataManager.getArtifactStore( key );
+            return clearPathProcessor( path, store, groups );
+        }
+        catch ( IndyDataException e )
+        {
+            logger.error( "Cannot clear paths for missing / inaccessible store: " + key, e );
+        }
+
+        return ()->0;
+    }
+
+    private Callable<Integer> clearPathProcessor( final String path, final ArtifactStore store, final Set<Group> groups )
+    {
+        return () -> {
+            logger.debug( "Got mergable transfer from diverged portion of membership: {}", path );
+
+            int cleared = clearPath( path, store, groups, false );
+            return cleared;
+        };
+    }
+
+    private int clearPath( String path, ArtifactStore origin, Set<Group> affectedGroups, boolean deleteOriginPath )
     {
         Logger logger = LoggerFactory.getLogger( getClass() );
-        final boolean isReadonlyHosted = storeDataManager.isReadonly( originKey );
-        paths.parallelStream()
-                 .forEach( p -> {
-                     if ( deleteOriginPath && !isReadonlyHosted )
-                     {
-                         try
-                         {
-                             delete( directContentAccess.getTransfer( originKey, p ) );
-                         }
-                         catch ( IndyWorkflowException e )
-                         {
-                             logger.error( String.format( "Failed to retrieve transfer for: %s in origin store: %s. Reason: %s", p, originKey, e.getMessage() ), e );
-                         }
-                     }
-
-                     affectedGroups.parallelStream().forEach( g->{
-                         try
-                         {
-                             Transfer gt = directContentAccess.getTransfer( g, p );
-                             delete( gt );
-                         }
-                         catch ( IndyWorkflowException e )
-                         {
-                             logger.error( String.format( "Failed to retrieve transfer for: %s in group: %s. Reason: %s", p, g.getName(), e.getMessage() ), e );
-                         }
-                     } );
-                 } );
+        final boolean isReadonlyHosted = storeDataManager.isReadonly( origin );
+        AtomicInteger cleared = new AtomicInteger( 0 );
```

Comment:
Not a big issue. Just not sure if an atomicInteger is really needed here, because it is a local variable and not executed in a concurrent environment in this local method.

Reply:
ok I know why this is needed. It is executed in a lamda....